src/shogun/machine/BaggingMachine.cpp

```diff
@@ -203,12 +208,26 @@ bool BaggingMachine::train_machine(std::shared_ptr<Features> data)
 
 #pragma omp critical
 		{
-		// get out of bag indexes
-		auto oob = get_oob_indices(idx);
-		m_oob_indices.push_back(oob);
 
-		// add trained machine to bag array
-		m_bags.push_back(c);
+			// get out of bag indexes
+			auto oob = get_oob_indices(idx);
+			m_oob_indices.push_back(oob);
+
+			// add trained machine to bag array
+			m_bags.push_back(c);
+
+			// observe some variables. The oob error is computed only when when
+			// we have all the variables set and there are observers attached.
+			this->observe<std::shared_ptr<Machine>>(
+			    i, "trained_machine", "Trained machine for this bag", c);
+
+			if (this->m_combination_rule && this->m_oob_evaluation_metric &&
+			    this->get_num_subscriptions() != 0)
+			{
+				auto oob_error = this->get_oob_error();
```

Comment:
question: you now always compute the oob error whenever there is an observer ...Is that good? I forgot how the specifying to be observed variables works, but can't the user somehow pick what she wants to observe? In that case this should only be computed if the user requested it

Reply:
So, the idea is that as long as the model has some observer attached, then it will output all the available observations. This means that the OOB error will be computed regardless of the user preferences.

It will be the observer job to filter the observations and keep just the ones that the user wants.

Reply:
At the moment, there is no way to tell an observable model to emit only certain observations.

Reply:
I don’t think that’s good and we should probably think about a way to avoid this. Passing (and ignoring) pointers to existing data is one thing, but expensive operations should only be executed if desired.
I could imagine that this could be done with a lambda function that is only executed if the observer actually wants (as in user instructed) to store the data. Maybe @gf712 has an idea he is the lambda master?

Reply:
I see the problem. One small issue here is that the model does not have any idea about which things the observers want to observe. I guess the solution could be emitting functions (lambdas) instead which can then be executed by the observers if they want to store that kind of data. Is this what you meant @karlnapf?

Reply:
yes, the model doesnt know. In general, it should *offer* the observer to observe something (which the observer then can or cannot do, depending on its settings). Passing a pointer is a good example for that. The observer can clone the underlying structure for example.
But for costly things, passing a lambda that would compute something expensive is a better choice imo, so yet.

Reply:
Okay, that makes sense! I will do some experimenting to try adding this lambda feature. I'm leaving this one as it is (or maybe we could merge it in a `feature` branch) and then I will open a new PR with the lambda observers.

Reply:
I say let's just keep this PR open, and then you could rebase it once the the lambda thing works. Alternatively, just remove the oob error stuff and we can merge everything else?